enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+OpenShift groups the user appears in, and it will either add to/remove from/create
+the necessary groups for the user. The user should only be removed from the groups
+that were created for that specific identity provider (as per the annotation
+from the [API section](#api)). If such a group ends up being empty, it should
+be removed if it contains the `oauth.openshift.io/generated: true` annotation.
+
+When a group that would otherwise be created on sync already exists, the
+oauth-server should only add the user among the group members and add an
+`oauth.openshift.io/idp.<idpname>` annotation to mark users of that IdP appear
+in that group.
+
+This implementation is not solving an issue when a user can be sourced from
+multiple identity providers when `mappingMethod: add` is configured for these
+providers. Such a user would get group memberships in all the groups from
+all matching identity providers. An alternative implementation will be
+discussed in the [Alternatives section](#alternatives).
+
+### Risks and Mitigations
+
+Given that the group synchronization only happens on logins, there is a time
+window during which the synchronization of groups between the identity provider
+and OpenShift is not guaranteed. This would typically be an issue when someone's
+group membership gets revoked.
+
+In order to force OpenShift group membership synchronization on group membership
+revocation on the identity provider side, one might:
+
+1. revoke all the OAuthAccessTokens for a given user by removing them, leveraging
+   a `userName` field selector, thus forcing the user to relogin:
+
+```bash
+       oc delete oauthaccesstoken --field-selector=userName="<username>"
+```
+
+2. remove the user from the OpenShift group manually
+
+Users can also be added to groups in the IdP, in which case they are more
+likely to willingly cause a resync by `oc login` themselves.
+
+## Design Details
+
+### Open Questions
+
+1. Do we care about mapping the groups to their source identity provider per
+   session (i.e. a user is only a member of the groups sourced from the IdP
+   that they logged in to + any OpenShift, non-synced groups)?
+
+### Test Plan
+
+The authentication operator already has a way to configure Keycloak and
+communicate with its API. Use that to add test users and groups to test:
+
+1. user can still log in when groups are configured but no group
+   claims are received
+2. on login, user gets added to a previously non-existent group when
+   the group appears in the configured claims
+3. on login, user gets added to an existing group that appears in the
+   retrieved group claims
+4. on login, user gets removed from any existing group that was created
+   for the given identity provider (per annotation) when the group
+   no longer appears among those retrieved from the OIDC
+
+
+### Follow-up Work
+
+To synchronize groups in a time-based manner, an additional controller in the
```

Comment:
Syncing groups from an ID Token could help here as well. We could always check groups upon user login.

Reply:
Syncing groups on user login is the intent of this enhancement :slightly_smiling_face: The interesting part is what happens when the user is already logged in since currently there is no communication with the 3rd party IdP during that time today (that is the <user login, OpenShift access token expires> time interval).

Reply:
But this is fine I guess. This is why you should use short-lived tokens in general. Upon a refresh (exchanging a Refresh Token for an Access Token) you should get a new ID Token as well. So the general remediation for this problem is to shorted the Access Token live time.

Reply:
We don't do any refreshes though. Also we'll have to consider the very likely different lifetimes of a 3rd-party refresh token and OpenShift OAuthAccessToken.

Reply:
Ok, understood. Thanks for clarification. 

So I'm just wondering - could we assume that a user needs to log out and in in order to refresh his roles? I'm trying to avoid any background synchronization mechanism in order to simplify the implementation.

Reply:
They only need to log in again, which would update groups for all their other sessions, too.

Reply:
Ok, thanks!